subsys/disk/disk_access_sdhc.c

```diff
@@ -0,0 +1,1001 @@
+		seed ^= (seed & 0xff) << 5;
+	}
+
+	return seed;
+}
+
+/* Traces card traffic for SYS_LOG_LEVEL_DEBUG */
+static int sdhc_trace(struct sdhc_data *data, int dir, int err,
+		      const u8_t *buf, int len)
+{
+#if SYS_LOG_LEVEL >= SYS_LOG_LEVEL_DEBUG
+	if (err != 0) {
+		printk("(err=%d)", err);
+		data->trace_dir = 0;
+	}
+
+	if (dir != data->trace_dir) {
+		data->trace_dir = dir;
+
+		printk("\n");
+
+		if (dir == 1) {
+			printk(">>");
+		} else if (dir == -1) {
+			printk("<<");
+		}
+	}
+
+	for (; len != 0; len--) {
+		printk(" %x", *buf++);
+	}
+#endif
+	return err;
+}
+
+/* Returns true if an error code is retryable at the disk layer */
+static bool sdhc_is_retryable(int err)
+{
+	switch (err) {
+	case 0:
+		return false;
+	case -EILSEQ:
+	case -EIO:
+	case -ETIMEDOUT:
+		return true;
+	default:
+		return false;
+	}
+}
+
+/* Maps a flag based error code into a Zephyr errno */
+static int sdhc_map_flags(const struct sdhc_flag_map *map, int flags)
+{
+	if (flags < 0) {
+		return flags;
+	}
+
+	for (; map->mask != 0; map++) {
+		if ((flags & map->mask) == map->mask) {
+			return -map->err;
+		}
+	}
+
+	return -map->err;
+}
+
+/* Converts disk status into an error code */
+static int sdhc_map_disk_status(int status)
+{
+	return sdhc_map_flags(sdhc_disk_status_flags, status);
+}
+
+/* Converts the R1 response flags into an error code */
+static int sdhc_map_r1_status(int status)
+{
+	return sdhc_map_flags(sdhc_r1_flags, status);
+}
+
+/* Converts an eary stage idle mode R1 code into an error code */
+static int sdhc_map_r1_idle_status(int status)
+{
+	if (status < 0) {
+		return status;
+	}
+	if (status == SDHC_R1_IDLE) {
+		return 0;
+	}
+	return sdhc_map_r1_status(status);
+}
+
+/* Converts the data block response flags into an error code */
+static int sdhc_map_data_status(int status)
+{
+	return sdhc_map_flags(sdhc_data_response_flags, status);
+}
+
+/* Initialises a retry helper */
+static void sdhc_retry_init(struct sdhc_retry *retry, u32_t timeout,
+			    u16_t sleep)
+{
+	retry->end = k_uptime_get_32() + timeout;
+	retry->tries = 0;
+	retry->sleep = sleep;
+}
+
+/* Called at the end of a retry loop.  Returns if the minimum try
+ * count and timeout has passed.  Delays/yields on retry.
+ */
+static bool sdhc_retry_ok(struct sdhc_retry *retry)
+{
+	s32_t remain = retry->end - k_uptime_get_32();
+
+	if (retry->tries < SDHC_MIN_TRIES) {
+		retry->tries++;
+		if (retry->sleep != 0) {
+			k_sleep(retry->sleep);
+		}
+		return true;
+	}
+
+	if (remain >= 0) {
+		if (retry->sleep > 0) {
+			k_sleep(retry->sleep);
+		} else {
+			k_yield();
+		}
+		return true;
+	}
+
+	return false;
+}
+
+/* Asserts or deasserts chip select */
+static void sdhc_set_cs(struct sdhc_data *data, int value)
+{
+	gpio_pin_write(data->cs, data->pin, value);
+}
+
+/* Receives a fixed number of bytes */
+static int sdhc_rx_bytes(struct sdhc_data *data, u8_t *buf, int len)
+{
+	struct spi_buf tx_bufs[] = {
+		{.buf = (u8_t *)sdhc_ones, .len = len},
```

Comment:
hm
{
.buf = ...,
.len = ...
};

would be nicer.

Reply:
Done globally.